src/clients/Deluge.ts

```diff
@@ -0,0 +1,260 @@
+			delugeRpcUrl
+		).unwrapOrThrow(
+			new CrossSeedError("delugeRpcUrl must be percent-encoded")
+		);
+		if (!password) {
+			throw new CrossSeedError(
+				"You need to define a password in the delugeRpcUrl. (e.g. http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		const response = await this.call("auth.login", [password], 0);
+		if (!response.result) {
+			throw new CrossSeedError(
+				`Reached Deluge, but failed to authenticate: ${href}`
+			);
+		}
+	}
+
+	/**
+	 * ensures authentication and sends JSON-RPC calls to deluge
+	 */
+	private async call(method: string, params: object, retries = 1) {
+		const { delugeRpcUrl } = getRuntimeConfig();
+		const { href } =
+			extractCredentialsFromUrl(delugeRpcUrl).unwrapOrThrow();
+		const headers = new Headers({ "Content-Type": "application/json" });
+		if (this.delugeCookie) headers.set("Cookie", this.delugeCookie);
+
+		let response: Response, json: DelugeResponse;
+		try {
+			response = await fetch(href, {
+				body: JSON.stringify({
+					...{ method: method, params: params },
+					id: Math.floor(Math.random() * 0x7fffffff),
+				}),
+				method: "POST",
+				headers: headers,
+			});
+		} catch (networkError) {
+			// @ts-expect-error needs es2022 target (tsconfig)
+			throw new Error(`Failed to connect to Deluge at ${href}`, {
+				cause: networkError,
+			});
+		}
+		try {
+			json = await response.json();
+		} catch (jsonParseError) {
+			throw new Error(
+				`Deluge method ${method} response was non-JSON ${jsonParseError}`
+			);
+		}
+		if (json?.error?.code === DelugeErrorCode.NO_AUTH && retries > 0) {
+			this.delugeCookie = null;
+			await this.authenticate();
+			if (this.delugeCookie) {
+				return this.call(method, params, 0);
+			} else {
+				throw new Error(
+					"Connection lost with Deluge. Reauthentication failed."
+				);
+			}
+		}
+		this.handleResponseHeaders(response.headers);
+		return json;
+	}
+
+	/**
+	 * parses the set-cookie header and updates stored value
+	 */
+	private handleResponseHeaders(headers: Headers) {
+		if (headers.has("Set-Cookie")) {
+			this.delugeCookie = headers.get("Set-Cookie").split(";")[0];
+		}
+	}
+
+	/**
+	 * checks enabled plugins for "Label"
+	 * returns true if successful.
+	 */
+	private async labelEnabled() {
+		const enabledLabels = await this.call("core.get_enabled_plugins", []);
+		return enabledLabels?.result?.includes?.("Label");
+	}
+
+	/**
+	 * if Label plugin is loaded, adds (if necessary)
+	 * and sets the label based on torrent hash.
+	 */
+	private async setLabel(infoHash: string, label: string): Promise<void> {
+		if (this.isLabelEnabled) {
+			const setResult = await this.call("label.set_torrent", [
+				infoHash,
+				label,
+			]);
+			if (setResult?.error?.code == DelugeErrorCode.RPC_FAIL) {
+				await this.call("label.add", [label]);
+				await this.call("label.set_torrent", [infoHash, label]);
+			}
+		}
+	}
+
+	/**
+	 * injects a torrent into deluge client
+	 */
+	async inject(
```

Comment:
can we wrap this whole function in a try/catch logging and returning FAILURE to make sure we don't crash the process?